Introduce multiple inheritance model

stevenhwu opened the pull request:
- This is relate to #54, first of many PR.
- Update `call.xmh` to take `--model`
- Model information are parsed by `InheritanceModel` class. No functional changed in RelationshipGraph Class yet (in next PR).
- Add `Gender` to `ped.h`, and vertex_gender to `graph.h`
- Gender information is passed throgh vertex (boost:property_map...vertex_gender_t> down to libries in RelationshipGraph class. It is stored in transition_t as well, but most likely this will be changed later. No functional change yet (in next PR).
- Check dad is male and mom is female.
- Rename variable name/class member from `pedigree` to `relationship_graph`. Finishing #85

stevenhwu committed:
Introduce multiple inheritance model

- Update `call.xmh` to take `--model`
- Model information are parsed by `InheritanceModel` class. No functional changed in RelationshipGraph Class yet (in next PR).
- Add `Gender` to `ped.h`, and vertex_gender to `graph.h`
- Gender information is passed throgh vertex (boost:property_map...vertex_gender_t> down to libries in RelationshipGraph class. It is stored in transition_t as well, but most likely this will be changed later. No functional change yet (in next PR).
- Check dad is male and mom is female.
- Rename variable name/class member from `pedigree` to `relationship_graph`

dng-jenkins commented:
Can one of the admins verify this patch?

reedacartwright commented:
Jenkins, okay to test

On Sep 23, 2016 01:35, "Steven Wu" <notifications@github.com> wrote:

>
>    - This is relate to #54
>    <https://github.com/denovogear/denovogear/issues/54>, first of many PR.
>    - Update call.xmh to take --model
>    - Model information are parsed by InheritanceModel class. No
>    functional changed in RelationshipGraph Class yet (in next PR).
>    - Add Gender to ped.h, and vertex_gender to graph.h
>    - Gender information is passed throgh vertex
>    (boost:property_map...vertex_gender_t> down to libries in
>    RelationshipGraph class. It is stored in transition_t as well, but most
>    likely this will be changed later. No functional change yet (in next PR).
>    - Check dad is male and mom is female.
>    - Rename variable name/class member from pedigree to relationship_graph.
>    Finishing #85 <https://github.com/denovogear/denovogear/issues/85>
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/denovogear/denovogear/pull/193
> Commit Summary
>
>    - Introduce multiple inheritance model
>
> File Changes
>
>    - *M* src/include/dng/find_mutations.h
>    <https://github.com/denovogear/denovogear/pull/193/files#diff-0> (5)
>    - *M* src/include/dng/graph.h
>    <https://github.com/denovogear/denovogear/pull/193/files#diff-1> (16)
>    - *A* src/include/dng/inheritance_model.h
>    <https://github.com/denovogear/denovogear/pull/193/files#diff-2> (66)
>    - *M* src/include/dng/io/ped.h
>    <https://github.com/denovogear/denovogear/pull/193/files#diff-3> (2)
>    - *M* src/include/dng/relationship_graph.h
>    <https://github.com/denovogear/denovogear/pull/193/files#diff-4> (45)
>    - *M* src/include/dng/task/call.xmh
>    <https://github.com/denovogear/denovogear/pull/193/files#diff-5> (4)
>    - *M* src/lib/CMakeLists.txt
>    <https://github.com/denovogear/denovogear/pull/193/files#diff-6> (11)
>    - *M* src/lib/find_mutations.cc
>    <https://github.com/denovogear/denovogear/pull/193/files#diff-7> (31)
>    - *A* src/lib/inheritance_model.cc
>    <https://github.com/denovogear/denovogear/pull/193/files#diff-8> (66)
>    - *M* src/lib/relationship_graph.cc
>    <https://github.com/denovogear/denovogear/pull/193/files#diff-9> (153)
>    - *M* src/lib/task/call.cc
>    <https://github.com/denovogear/denovogear/pull/193/files#diff-10> (21)
>
> Patch Links:
>
>    - https://github.com/denovogear/denovogear/pull/193.patch
>    - https://github.com/denovogear/denovogear/pull/193.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/denovogear/denovogear/pull/193>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGOHsRGSWm9LhlSs8G4Z6foate45hbIks5qs482gaJpZM4KEu-8>
> .
>

stevenhwu committed:
Add missing header

- add `#include <stdexcept>` for FreeBSD9 and CentOS

stevenhwu committed:
Merge branch 'develop' into feature/multi_inheritance_PR1

reedacartwright commented:
@stevenhwu, conflicts need to be now fixed.

stevenhwu committed:
Fixed merge conflict after PR #179

reedacartwright commented on src/include/dng/graph.h:
```diff
@@ -20,20 +20,25 @@
 BOOST_INSTALL_PROPERTY(edge, length);
 BOOST_INSTALL_PROPERTY(edge, type);
+BOOST_INSTALL_PROPERTY(vertex, gender);
 BOOST_INSTALL_PROPERTY(vertex, label);
```
This property should be `sex` not `gender`

reedacartwright commented on src/include/dng/graph.h:
```diff
@@ -46,14 +51,15 @@ enum struct EdgeType : std::size_t {
 };
 
-typedef boost::property<boost::vertex_group_t, std::size_t> VertexGroupProp;
+typedef boost::property<boost::vertex_gender_t, dng::io::Pedigree::Gender> VertexGenderProp;
```
`vertex_sex_t`

reedacartwright commented on src/include/dng/inheritance_model.h:
```diff
@@ -0,0 +1,66 @@
+//        maternal (transmitted by mother to child)
+//        paternal (transmitter by father to child)
+//        mitochondria  (transmitted by mother to child)
+};
```
Should we move this comment to the same line?

reedacartwright commented on src/include/dng/inheritance_model.h:
```diff
@@ -0,0 +1,66 @@
+enum class InheritancePattern : int {
+    AUTOSOMAL = 0,
+    DEFAULT = 0,
+    MATERNAL = 1,
```
Put DEFAULT above AUTOSOMAL and MITOCHONDRIAL with MATERNAL

reedacartwright commented on src/include/dng/io/ped.h:
```diff
@@ -57,7 +57,7 @@ class Pedigree {
         std::size_t mom;
-        Gender sex;
+        Gender gender;
         std::string sample_tree;
```
Sex sex

reedacartwright commented on src/include/dng/relationship_graph.h:
```diff
@@ -47,31 +48,12 @@ class RelationshipGraph {
     typedef boost::property_map<Graph, boost::vertex_group_t>::type PropVertexGroup;
     typedef boost::property_map<Graph, boost::vertex_index_t>::type PropVertexIndex;
+    typedef boost::property_map<Graph, boost::vertex_gender_t>::type PropVertexGender;
 
```
vertex_sex_t

reedacartwright commented on src/include/dng/relationship_graph.h:
```diff
@@ -96,26 +78,28 @@ class RelationshipGraph {
         double length1;
         double length2;
+        dng::io::Pedigree::Gender gender;
     };
```
Sex

reedacartwright commented on src/include/dng/task/call.xmh:
```diff
@@ -54,8 +54,8 @@ XM((rgtag), , "combine read groups using @RG tags, e.g. ID, SM, LB, or DS.",
-
-
+XM((model), (M), "Inheritance model, currently only support autosomal", std::string, "Autosomal")
+//TODO(SW): Change help text later
```
default should be "autosomal" (no capital letter)

reedacartwright commented on src/lib/inheritance_model.cc:
```diff
@@ -0,0 +1,67 @@
+}
+
+void dng::InheritanceModel::parse_model(std::string &model_string) {
+
```
You want to use `dng::utility::key_switch` or `dng::utility::key_switch_tuple` here.

reedacartwright commented on src/lib/relationship_graph.cc:
```diff
@@ -63,7 +75,7 @@ bool dng::RelationshipGraph::Construct(const io::Pedigree &pedigree,
     auto labels = get(boost::vertex_label, pedigree_graph);
-
+    auto gender  = get(boost::vertex_gender, pedigree_graph);
 
```
sex

reedacartwright commented on src/lib/relationship_graph.cc:
```diff
@@ -347,10 +345,24 @@ std::vector<std::string> dng::RelationshipGraph::BCFHeaderLines() const {
 
     auto labels = get(boost::vertex_label, pedigree_graph);
+    auto gender  = get(boost::vertex_gender, pedigree_graph);
     labels[0] = DNG_GL_PREFIX "unknown";
```
sex

reedacartwright commented on src/lib/relationship_graph.cc:
```diff
@@ -361,9 +373,11 @@ void dng::RelationshipGraph::ParseIoPedigree(dng::Graph &pedigree_graph,
         vertex_t mom = row.mom;
-
+        gender[child] = row.gender;
 #ifdef DEBUG_RGRAPH
```
sex

reedacartwright commented on src/lib/relationship_graph.cc:
```diff
@@ -361,9 +373,11 @@ void dng::RelationshipGraph::ParseIoPedigree(dng::Graph &pedigree_graph,
+        std::cout << "\n==START\n===Child_Dad_MoM_Gender: " << child << "\t"
+            << dad << "\t" << mom << "\t"
+            << (int) row.gender << std::endl;
 #endif
```
sex

reedacartwright commented on src/lib/relationship_graph.cc:
```diff
@@ -375,6 +389,12 @@ void dng::RelationshipGraph::ParseIoPedigree(dng::Graph &pedigree_graph,
+                    "Error: Gender error, dad!=Male || mom!=Female");
+        }
+
         // check to see if mom and dad have been seen before
```
sex

reedacartwright commented on src/lib/relationship_graph.cc:
```diff
@@ -383,6 +403,10 @@ void dng::RelationshipGraph::ParseIoPedigree(dng::Graph &pedigree_graph,
 #ifdef DEBUG_RGRAPH
     std::cout << "===after spoesal: V E:" << num_vertices(pedigree_graph) << "\t" << num_edges(pedigree_graph) << std::endl;
+    int gender2 = static_cast<int>( gender[child] );
+        std::cout << "\n===Child_Dad_MoM_gender: " << child << "\t" << dad
```
sex

reedacartwright commented on src/lib/relationship_graph.cc:
```diff
@@ -383,6 +403,10 @@ void dng::RelationshipGraph::ParseIoPedigree(dng::Graph &pedigree_graph,
+    int gender2 = static_cast<int>( gender[child] );
+        std::cout << "\n===Child_Dad_MoM_gender: " << child << "\t" << dad
+                << "\t" << mom << "\tgender: " << (int) gender[child] << "\t" << gender2
+                << std::endl;
```
sex

reedacartwright commented on src/lib/relationship_graph.cc:
```diff
@@ -394,18 +418,27 @@ void dng::RelationshipGraph::ParseIoPedigree(dng::Graph &pedigree_graph,
+        for (int i = current_index; i < num_vertices(pedigree_graph); ++i) {
+            std::cout << i << "\t" << num_vertices(pedigree_graph) << std::endl;
+            gender[i] = gender[child];
+        }
```
sex

reedacartwright commented on src/lib/relationship_graph.cc:
```diff
@@ -394,18 +418,27 @@ void dng::RelationshipGraph::ParseIoPedigree(dng::Graph &pedigree_graph,
             vertex_t v = add_vertex(DNG_SM_PREFIX + pedigree.name(child), pedigree_graph);
             add_edge(child, v, {EdgeType::Mitotic, 1.0f}, pedigree_graph);
+            gender[v] = gender[child];
 
```
sex

reedacartwright commented on src/lib/relationship_graph.cc:
```diff
@@ -453,17 +486,20 @@ void dng::RelationshipGraph::ConnectSomaticToLibraries(
         const PropVertexLabel &labels) {
 
+    auto gender  = get(boost::vertex_gender, pedigree_graph);
     const size_t STRLEN_DNG_SM_PREFIX = strlen(DNG_SM_PREFIX);
```
sex

reedacartwright commented on src/lib/relationship_graph.cc:
```diff
@@ -453,17 +486,20 @@ void dng::RelationshipGraph::ConnectSomaticToLibraries(
         for (; r.first != r.second; ++r.first) {
             vertex_t w = first_library_ + rg::index(rgs.libraries(), r.first->library);
+            gender[w] = gender[v];
             if (!edge(v, w, pedigree_graph).second) {
```
sex

reedacartwright commented on src/lib/relationship_graph.cc:
```diff
@@ -519,6 +553,9 @@ void dng::RelationshipGraph::SimplifyPedigree(dng::Graph &pedigree_graph) {
             // this node has no descendants
             clear_vertex(v, pedigree_graph);
+#ifdef DEBUG_RGRAPH
+    std::cout << "Remove chiled==0:" << v << std::endl;
```
typo

reedacartwright commented on src/lib/relationship_graph.cc:
```diff
@@ -588,14 +625,16 @@ void dng::RelationshipGraph::UpdateLabelsNodeIds(dng::Graph &pedigree_graph,
 
 
 #ifdef DEBUG_RGRAPH
+    auto gender  = get(boost::vertex_gender, pedigree_graph);
```
sex

reedacartwright commented on src/lib/relationship_graph.cc:
```diff
@@ -588,14 +625,16 @@ void dng::RelationshipGraph::UpdateLabelsNodeIds(dng::Graph &pedigree_graph,
     std::cout << "node_ids:" << std::endl;
     for (int j = 0; j < node_ids.size(); ++j) {
-        std::cout << j << " -> " << node_ids[j] << std::endl;
+        std::cout << j << " -> " << node_ids[j] << "\tgender:"
```
sex

reedacartwright commented on src/lib/relationship_graph.cc:
```diff
@@ -702,12 +740,25 @@ void dng::RelationshipGraph::CreatePeelingOps(
 
     auto edge_types = get(boost::edge_type, pedigree_graph);
     auto lengths = get(boost::edge_length, pedigree_graph);
+    auto gender  = get(boost::vertex_gender, pedigree_graph);
```
sex

reedacartwright commented on src/lib/relationship_graph.cc:
```diff
@@ -702,12 +740,25 @@ void dng::RelationshipGraph::CreatePeelingOps(
+    auto gender  = get(boost::vertex_gender, pedigree_graph);
 
     ResetFamilyInfo();
+    for (std::size_t i = first_founder_; i < first_nonfounder_; ++i) {
```
I'm pretty sure this can be optimized.

stevenhwu replied:
I think the hack k==i+1 is true all the time.
But, let's redo this part later, after all inheritance models are implemented.
Added PR_NOTE and test/refactor this later.

reedacartwright commented on src/lib/relationship_graph.cc:
```diff
@@ -749,12 +800,14 @@ void dng::RelationshipGraph::CreatePeelingOps(
-            transitions_[child] = {tt, parent, static_cast<size_t>(-1), lengths[*pos], 0};
+            transitions_[child] = {tt, parent, static_cast<size_t>(-1),
+                    lengths[*pos], 0, gender[child_index]};//TODO(SW): Double check the index is correct
             family_members_.push_back({parent, child});
```
sex

reedacartwright commented on src/lib/relationship_graph.cc:
```diff
@@ -852,29 +906,40 @@ void dng::RelationshipGraph::ResetFamilyInfo(){
 //    typedef property_map<Graph, vertex_index_t>::type IndexMap;
+    int verbose_level = 2;
     PropVertexIndex index = get(boost::vertex_index, pedigree_graph);
+    auto gender = get(boost::vertex_gender, pedigree_graph);
```
sex

reedacartwright commented on src/lib/relationship_graph.cc:
```diff
@@ -852,29 +906,40 @@ void dng::RelationshipGraph::ResetFamilyInfo(){
+    if (verbose_level > 1) {
+        boost::graph_traits<Graph>::vertex_iterator vi, vi_end;
+        for (boost::tie(vi, vi_end) = vertices(pedigree_graph); vi != vi_end; ++vi) {
+            std::cout << "Vertex_Gender:" << *vi << "_" << (int) gender[*vi] << ". ";
```
sex

reedacartwright reviewed (changes requested):
- `Gender` should be changed to `sex` throughout the request.
  - The parsing of the inheritance model from a string can be simplified as well using utility functions.
  - The function to create transitions_ probably doesn't need to use std::find and std::distance.

stevenhwu committed:
Review the request changed

- [x] `Gender` should be changed to sex throughout the request.
- [x] The parsing of the inheritance model from a string can be simplified as well using utility functions.
- [ ] The function to create transitions_ probably doesn't need to use std::find and std::distance.
Response: I think the hack k==i+1 is true all the time.
But, let's redo this part later, after all inheritance models are implemented.
Added this in PR_NOTE and test/refactor this later.

reedacartwright commented on src/include/dng/inheritance_model.h:
```diff
@@ -54,6 +49,25 @@ class InheritanceModel{
 
+
+    const std::pair<std::string, InheritancePattern> INHERITANCE_KEYS[13] = {
+        {"DEFAULT", InheritancePattern::AUTOSOMAL},
```
Probably don't need to support "DEFAULT" as a key, but if we do, it should be linked to DEFAULT, not AUTOSOMAL.

reedacartwright commented on src/lib/inheritance_model.cc:
```diff
@@ -29,37 +32,15 @@ dng::InheritanceModel::InheritanceModel() {
-            break;
+    auto match = dng::utility::key_switch_tuple(model_string, INHERITANCE_KEYS,
+            std::pair<std::string, InheritancePattern>{"UNMATCHED", InheritancePattern::DEFAULT});
+    if ("UNMATCHED" == match.first ){
```
This can be simplified if there existed an InheritancePattern::UNKOWN key (set to value -1), then we could do an integer comparison instead of a string comparison.

And we should toss an exception if it is not matched.

reedacartwright replied:
Also toss an exception if model_string is empty, because otherwise it will match the first element, which we might not want to do.

reedacartwright commented on src/lib/inheritance_model.cc:
```diff
@@ -29,37 +32,15 @@ dng::InheritanceModel::InheritanceModel() {
-            break;
+    auto match = dng::utility::key_switch_tuple(model_string, INHERITANCE_KEYS,
+            std::pair<std::string, InheritancePattern>{"UNMATCHED", InheritancePattern::DEFAULT});
+    if ("UNMATCHED" == match.first ){
```
Also toss an exception if model_string is empty, because otherwise it will match the first element, which we might not want to do.

stevenhwu committed:
Update for review round 2

- Remove InheritancePattern::Default
- Add InheritancePattern::UNKOWN and throws error for --model "", --model " ", and all other unmatched one
